### PR TITLE
Add checked div by weight to amount

### PR DESCRIPTION
**Pull request by yancyribbens**
Adds the checked variant of `amount / weight`.  I also added a test to the non-checked version for comparison so the reviewer knows they compute the same way (integer division rounded down).

Also added assertion to `checked_weight_mul test` showing the results are rounded up.

### CODE REVIEW

**Comment by jaoleal** [comment]
Check `just lint` and `just format`.

Successfully ran test and build locally

File: units/src/amount.rs [hunk 1]
```diff
@@ -5,6 +5,8 @@
 //! This module mainly introduces the [`Amount`] and [`SignedAmount`] types.
 //! We refer to the documentation on the types for more information.
 
+use crate::{Weight, FeeRate};
```
Comment (jamillambert) [inline units/src/amount.rs:8]: Should be 
```rust
use crate::fee_rate::FeeRate;
use crate::weight::Weight;
```

Comment (jamillambert) [reply units/src/amount.rs:9]: Yeah the feature gate fixed the issue.  All looks good now.

**Comment by apoelstra** [comment]
We should be rounding up, not down.

**Comment by yancyribbens** [comment]
> We should be rounding up, not down.

I thought that might be the case.  Probably should have inquired about changing it first.

**Comment by yancyribbens** [comment]
> We should be rounding up, not down.

I don't disagree that it makes sense to round up, although I think that would be confusing if the behavior of amount / weight is rounding down while here in this PR we are rounding up.

**Comment by apoelstra** [comment]
We should be sure to document the difference, but I don't think it's confusing at all.

**Comment by yancyribbens** [comment]
> We should be sure to document the difference, but I don't think it's confusing at all.

For my particular use case, I don't care how this rounds.  Adding some docs is sensible, I just don't want the mob to come after me if someone disagrees.

**Comment by yancyribbens** [comment]
Updated to round up https://github.com/rust-bitcoin/rust-bitcoin/pull/3430/commits/45d65393b3af70708f21a18a49b377c2d49fe927

**Comment by yancyribbens** [comment]
I added a `checked_add` by 1 here https://github.com/rust-bitcoin/rust-bitcoin/pull/3430/commits/3e73807c70c85be2d684846cb7fb83633a263323 although in practice I don't think it's possible that it will overflow.

File: units/src/amount.rs [hunk 2]
```diff
@@ -1039,6 +1042,29 @@ impl Amount {
...
+    /// [`None`] is returned if an overflow occurred.
+    #[cfg(feature = "alloc")]
+    pub fn checked_div_by_weight(self, rhs: Weight) -> Option<FeeRate> {
+        let sats = self.0.checked_mul(1000)?;
+        let wu = rhs.to_wu();
+
+        let fee_rate = sats.checked_div(wu)?;
+
+        let result = if sats.checked_rem(wu)? != 0 {
+            fee_rate.checked_add(1)?
```
Comment (apoelstra) [inline units/src/amount.rs:1060]: In 3e73807c70c85be2d684846cb7fb83633a263323:

Please just do this the normal way, by adding `wu - 1` before dividing by `wu`. This is much faster, easier to read, and less code.

Comment (yancyribbens) [reply units/src/amount.rs]: Thanks, I updated it.  I had actually tried this way originally, trying to follow how `checked_weight_by_mul` works and add 999 then divide by 1000 but I kept running into incorrect results.  And this way conceptually made sense to me.  Anyway, your suggestions of add wu -1 and then divide by wu also gives the correct results for my tests and if this is standard for the project that works for me.

**Comment by apoelstra** [comment]
utack 8d074cb592e31597a19382334b5f1ebbc0a2a567

I am a tiiny bit tempted to say something about how if `Amount` is close to `u64::MAX/1000` then the division might fail due to the `checked_add` or the `checked_mul`, but that seems overly verbose because no real user will ever hit this. (I am also tempted to promote these to panics, but since it's possible to construct these sorts of obscene values, we ought to handle them gracefully.)

**Comment by yancyribbens** [comment]
> I am a tiiny bit tempted to say something about how if Amount is close to u64::MAX/1000 then the division might fail due to the checked_add or the checked_mul, but that seems overly verbose because no real user will ever hit this. 

That's true.  I actually hit this during some proptests because of the multiply by `999` which is why I wanted to add this checked variant.  At least this way, it can be caught and handled instead of a panic.

> (I am also tempted to promote these to panics, but since it's possible to construct these sorts of obscene values, we ought to handle them gracefully.)

there's still the unchecked version of this (which rounds down) if the desired behavior is to panic..

**Comment by yancyribbens** [comment]
In the proptest, I'm calculating what the max fee_rate would be for a theoretical pool of utxos in the proptest.

**Comment by yancyribbensmobile** [comment]
Oops, this should probably use `checked_sub` instead of `wu - 1` which would catch the case were weight is zero.  Will update when I'm back online.

**Comment by apoelstra** [comment]
Oops, good catch. Should just unconditionally return 0 in this case, no need to use checked_sub.

**Comment by yancyribbens** [comment]
> there's still the unchecked version of this (which rounds down) if the desired behavior is to panic..

Also, this isn't correct since release mode will overflow if I recall.  I forget that each mode handles this differently.

**Comment by yancyribbens** [comment]
> Should just unconditionally return 0 in this case, no need to use checked_sub.

div by zero is what's being guarded by `checked_div(wu)?`.  Would that mean if weight is zero, just return zero in this case as well?  It seems like using checked in all cases is simpler, especially to document, although I'm asking because I don't know.

**Comment by apoelstra** [comment]
Your existing code will panic if you pass in 0. If you use `checked_sub` then it will return `None`. The correct result is `Some(0)`.

**Comment by yancyribbens** [comment]
> Your existing code will panic if you pass in 0. If you use checked_sub then it will return None. The correct result is Some(0).

see added assertion:
```rust
let fee_rate = Amount::ONE_SAT.checked_div_by_weight(Weight::ZERO).unwrap();
assert_eq!(fee_rate, FeeRate::ZERO);
```

**Comment by apoelstra** [comment]
Ugh. Division by 0 should return `None`. I thought we were talking about division where `self` was 0. I apologize.

**Comment by yancyribbens** [comment]
> Ugh. Division by 0 should return None. I thought we were talking about division where self was 0. I apologize.

I'm glad to hear that.  I had felt conflicted by this change since any calculator will treat div by zero as err no zero.

**Review by apoelstra — APPROVED** [review]
ACK 016cb242c4e80e6d827a38f3a8545df8d0ae9905 successfully ran local tests

File: units/src/amount.rs [hunk 3]
```diff
@@ -2219,6 +2238,31 @@ mod tests {
...
+            .checked_div_by_weight(weight)
+            .unwrap();
+        // 1 sats * 1,000 wu = 1 sats/kwu
+        assert_eq!(fee_rate, FeeRate::from_sat_per_kwu(1));
+
+        let weight = Weight::from_wu(381);
+        let fee_rate = Amount::from_sat(329)
+            .checked_div_by_weight(weight)
+            .unwrap();
+        // 329 sats * 381 wu = 863.5 sats/kwu
```
Comment (tcharding) [inline units/src/amount.rs:2255]: Is this `*` sign meant to be a `/` sign? (one more below too)

Comment (yancyribbens) [reply units/src/amount.rs]: Yes! it is supposed to be /.  Updated, thanks.

**Review by tcharding — APPROVED** [review]
ACK a0c58a4a8b4244d7c541906c61d1343dd6acdccd

**Review by apoelstra — APPROVED** [review]
ACK a0c58a4a8b4244d7c541906c61d1343dd6acdccd successfully ran local tests

**Pull request merged**